### PR TITLE
Add WAF ACL to Cognito User Pool

### PR DESCRIPTION
**Pull request by noah-paige**
### Feature or Bugfix

- Feature

### Detail
- Add WAF ACL to Cognito

### Relates
- https://github.com/data-dot-all/dataall/issues/902

### Security
Please answer the questions below briefly where applicable, or write `N/A`. Based on
[OWASP 10](https://owasp.org/Top10/en/).

- Does this PR introduce or modify any input fields or queries - this includes
fetching data from storage outside the application (e.g. a database, an S3 bucket)?
  - Is the input sanitized?
  - What precautions are you taking before deserializing the data you consume?
  - Is injection prevented by parametrizing queries?
  - Have you ensured no `eval` or similar functions are used?
- Does this PR introduce any functionality or component that requires authorization?
  - How have you ensured it respects the existing AuthN/AuthZ mechanisms?
  - Are you logging failed auth attempts?
- Are you using or adding any cryptographic features?
  - Do you use a standard proven implementations?
  - Are the used keys controlled by the customer? Where are they stored?
- Are you introducing any new policies/roles/users?
  - Have you used the least-privilege principle? How?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: deploy/stacks/cognito.py [hunk 1]
```diff
@@ -50,6 +53,41 @@ def __init__(
...
             advanced_security_mode='ENFORCED'
         )
+
+        # Create IP set if IP filtering enabled in CDK.json
+        ip_set_regional = None
+        if custom_waf_rules and custom_waf_rules.get('allowed_ip_list'):
+            ip_set_regional = wafv2.CfnIPSet(
+                self,
+                'DataallRegionalIPSet-Cognito',
+                name=f'{resource_prefix}-{envname}-ipset-regional',
```
Comment (dlpzx) [inline deploy/stacks/cognito.py:63]: The name of the IP set is the same one as for the WAF of API GW in `lambda_api.py`. This will cause `WAFDuplicateItemException: AWS WAF couldn’t perform the operation because some resource in your request is a duplicate of an existing one.` when `allowed_ip_list` is defined

Comment (dlpzx) [reply deploy/stacks/cognito.py:63]: We can create 2 different IP sets (as we are trying to do now) or we can think of how to merge it in one. I personally see the value fo having 2 separate IP sets that allows this feature to be extended and restrict IP addresses differently for Cognito and for API GW

File: deploy/stacks/waf_rules.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+                sampled_requests_enabled=True,
+                cloud_watch_metrics_enabled=True,
+                metric_name='AWS-AWSManagedRulesKnownBadInputsRuleSet',
+            ),
+            priority=priority,
+            override_action=wafv2.CfnWebACL.OverrideActionProperty(none={}),
+        )
+    )
+    priority += 1
+    waf_rules.append(
```
Comment (dlpzx) [inline deploy/stacks/waf_rules.py:127]: I have been searching for recommended WAF rules for Cognito in particular. Because it is a recent launch the resources are a bit limited, but in general geo-fencing, IP-allow-list, rate limits and the typical AWSManagedRulesCommonRuleSet appear in demos and blogs.
From a cost perspective, WAF currently adds $23/month to the data.all bill. But if you see the docs: https://aws.amazon.com/waf/pricing/ going over 1500 WCUs affects the costs slightly `The total WCUs for a web ACL can't exceed 5000. Using over 1500 WCUs affects your costs.` we are currently at 1427.

In short:
- there are no additional WAF rules to add at the moment
- we might remove the SQL or the linux rules, but it is not a big deal because the costs are very low. wdyt?

**Review by dlpzx — CHANGES_REQUESTED** [review]
I will test with custom_waf_rules once the issue of the names of the IP sets is resolved

File: deploy/stacks/cognito.py (hunk 1 above)
Comment (noah-paige) [reply deploy/stacks/cognito.py:63]: good catch - updating the PR now

File: deploy/stacks/waf_rules.py (hunk 2 above)
Comment (noah-paige) [reply deploy/stacks/waf_rules.py:127]: Agreed - I could not find any prescriptive guidance on best practice WAF rules for Cognito. I think we are okay to keep rules the same for now and if needed break it out so each WebACL has its own unique rule set at a later point

**Comment by dlpzx** [comment]
Overall looks good! Testing in AWS with custom_waf_rules:

- [ ] CICD Pipeline Succeeds
- [ ] Added WAF Rules to Cognito User Pool (no updates to Cloudfront or API GW)
- [ ] Log In, Log Out, and Access Data.all UI

File: deploy/stacks/cognito.py [hunk 3]
```diff
@@ -26,6 +28,7 @@ def __init__(
         vpc=None,
         prod_sizing=False,
         internet_facing=True,
+        custom_waf_rules=None,
```
Comment (dlpzx) [inline deploy/stacks/cognito.py:31]: I have been tracking down this parameter and it seems unused upstream. It is not defined neither in the backend stack nor in the cdk.json file. Should we fix it in this PR or should we open a separate issue for it?

Comment (dlpzx) [inline deploy/stacks/cognito.py:31]: I have been tracking this parameter upstream and I cannot see it neither in the `cdk.json` nor in the backend stack. Should we re-enable it and add it to the cdk.json template and docs? In such case, maybe it is worth opening a separate PR. wdyt?

**Comment by dlpzx** [comment]
I created another issue for the `custom_waf_rules` : https://github.com/data-dot-all/dataall/issues/995
